File: Uncompress.c
/*
 * Copyright (c) 2009 Murtaza Munaim
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#include <stdio.h>
#include <math.h>
#include <sys/stat.h>
#include <sys/types.h>
#define MAX_ARGS 3
#define COMPRESS_SIZE 7
#define UNCOMPRESS_SIZE 8
#define SHIFT 8
#define MSB_CLEAR 127

long long ByteArrayToLongLong (unsigned char* arr, int bytes, int type);
double MyCeil(double input);
int MyPow(int base , int exponent);

int main(int argc, char** argv) {
   FILE *in, *out;
   long long origSize = 0;
   int ndx = 0, evenSize = 0, ctr = 0, endVal = 0;
   unsigned char byte, *decompress, readSize[UNCOMPRESS_SIZE];
   char *compress, byteBlock[UNCOMPRESS_SIZE];
   struct stat inStat, outStat;
   
   if (argc != MAX_ARGS) {
      printf("Usage: Compress fromFile toFile\n");
      exit(0);
   }   
   if (stat(argv[1], &inStat)) {
      printf("Input file %s doesn't exist or is unreadable.\n", argv[1]);
      exit(0);
   }
   if (!stat(argv[2], &outStat)) {
      printf("Cannot create %s, perhaps because", argv[2]);
      printf(" it already exists.\n");
      exit(0);
   }
   
   in = fopen(argv[1], "rb");
   out = fopen(argv[2], "w");  
   for (ndx = 0; ndx < UNCOMPRESS_SIZE; ndx++)
      fread(readSize + ndx, 1, 1, in);    
   origSize = ByteArrayToLongLong (readSize, UNCOMPRESS_SIZE, 1); 
   endVal = (int) MyCeil( (double)((origSize * (COMPRESS_SIZE * 1.0))
    / (UNCOMPRESS_SIZE * 1.0)));
   
   fseek(in, 0, SEEK_END);
   evenSize = ftell(in);
   fseek(in, 0, SEEK_SET);   
   if (endVal != (evenSize - UNCOMPRESS_SIZE)) {
      printf("Incorrect file length in compressed data.\n");
      exit(0);
   }
      
   while ((evenSize % COMPRESS_SIZE) != 0)
      evenSize++;  
   compress = (char*) malloc(evenSize * sizeof(char));
   memset(compress, 0, evenSize);
   decompress = (unsigned char*) 
    malloc((evenSize * UNCOMPRESS_SIZE)/ COMPRESS_SIZE);
 
   fseek(in, UNCOMPRESS_SIZE, SEEK_SET);
   for (ndx = 0; ndx < endVal; ndx++) {
      if (!feof(in))
         compress[ndx] = (char)fgetc(in); 
   }
   
   for (ndx = 0; ndx < evenSize ; ndx += COMPRESS_SIZE) {
      memset(byteBlock, 0, UNCOMPRESS_SIZE);
      endVal = ndx;
      ctr = 0;
      for (; ctr < COMPRESS_SIZE; ctr++, endVal++)
         byteBlock[ctr] = compress[endVal];
      byteBlock[COMPRESS_SIZE] = 0;
      
      for (ctr = COMPRESS_SIZE; ctr > 0; ctr--)
			byteBlock[ctr] = (unsigned char) (((byteBlock[ctr] >> (ctr + 1)) 
			 & (int)(MyPow(2, COMPRESS_SIZE - ctr) - 1)) 
			 | (byteBlock[ctr - 1] << (COMPRESS_SIZE - ctr))
			 & MSB_CLEAR );
      byteBlock[0] = (unsigned char) (byteBlock[0] >> 1 & MSB_CLEAR);
      
      endVal = ((ndx *UNCOMPRESS_SIZE) / COMPRESS_SIZE);
      ctr = 0;
      while (ctr < UNCOMPRESS_SIZE)
         decompress[endVal++] = byteBlock[ctr++];
   }
   
   for (ndx = 0; ndx < origSize; ndx++)
      fwrite(decompress + ndx, sizeof(unsigned char), 1, out);
   
   fclose(in);
   fclose(out);
}    

long long ByteArrayToLongLong (unsigned char* arr, int bytes, int type) { 
   long long num = (long long) 0, tmp;
   int ndx;
   for (ndx = 0; ndx < bytes; ndx++)
   {
      tmp = (long long) 0;
      tmp = arr[bytes - ndx - 1];
      num |= (tmp << ((ndx & COMPRESS_SIZE) << MAX_ARGS));
   }
   return num;
}

double MyCeil(double input) {
   double returns; 
   char *str;
   int ndx;
   ndx = snprintf(0, 0, "%.0f", input);
   str =(char *) malloc(ndx + 1); 
   sprintf(str, "%.0f", input); 
   sscanf(str, "%lf", &returns); 
   free(str);
   return input > returns ? returns + 1.0 : returns; 
}

int MyPow(int base , int exponent) {
   if (exponent == 0)
      return 1;
   else
      return (base << (exponent - 1));
}

File: Uncompress.java
/*
 * Copyright (c) 2009 Murtaza Munaim
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

import java.lang.Math;
import java.io.IOException;
import java.util.ArrayList;
import java.io.DataOutputStream;
import java.io.FileOutputStream;
import java.io.File;
import java.io.DataInputStream;
import java.io.FileInputStream;
import java.io.EOFException;

public class Uncompress {
	public static final int COMPRESS_SIZE = 7;
	public static final int UNCOMPRESS_SIZE = 8;
	public static final byte MSB_CLEAR = 127;
	
	public static void main (String[] args) {
		ArrayList<Byte> input = new ArrayList<Byte>();
		DataInputStream in;
		DataOutputStream out;
		long size = -1;
		int ndx = 0, ctr = 0;
		Byte [] decompress, compress, byteBlock = new Byte[8];

		if (args.length != 2) {
			System.out.println("Usage: Uncompress fromFile toFile");
			System.exit(0);
		}

		try {
			File inFile = new File(args[0]), outFile = new File(args[1]);

			if (outFile.exists()) {
				System.out.println("Cannot create " + outFile.getName() + 
				 ", perhaps because it already exists.");
				System.exit(0);
			}
			if (!inFile.canRead()) {
				System.out.println("Input file " + inFile.getName() + 
				 " doesn't exist or is unreadable.");
				System.exit(0);
			}

			out = new DataOutputStream(new FileOutputStream(outFile));
			in = new DataInputStream(new FileInputStream(inFile));

			size = in.readLong();

			try {
				for (ndx = 0; ndx < Math.ceil((size * (COMPRESS_SIZE * 1.0)) 
				 / (UNCOMPRESS_SIZE * 1.0)); ndx++) {
					try {
						input.add(in.readByte());
					}
					catch (EOFException eof) {
						System.out.println("Incorrect file length" +
						 " in compressed data.");
						break;
					}
				}
				in.readByte();
				System.out.println("Incorrect file length" + 
				 " in compressed data.");
				return;
			}
			catch (EOFException eof) {}

			catch (IOException e) {
				System.out.println ( "IO Exception =: " + e );
			}

			while (input.size() % COMPRESS_SIZE != 0)
				input.add((byte)(0));

			compress = new Byte[input.size()];
			decompress = new Byte[(input.size() *
			 UNCOMPRESS_SIZE) / COMPRESS_SIZE];
			compress = input.toArray(compress);

			for (ndx = 0; ndx < compress.length; ndx += COMPRESS_SIZE) {
				System.arraycopy(compress, ndx, byteBlock, 0, COMPRESS_SIZE);
				byteBlock[COMPRESS_SIZE] = 0;
				for (ctr = COMPRESS_SIZE; ctr > 0; ctr--)
					byteBlock[ctr] = (byte) ( ((byteBlock[ctr] >>> (ctr + 1)) 
					 & (int)(Math.pow(2, COMPRESS_SIZE - ctr) - 1)) 
					 | (byteBlock[ctr - 1] << (COMPRESS_SIZE-ctr))
					 & MSB_CLEAR );

				byteBlock[0] = (byte)(byteBlock[0]>>>1 & MSB_CLEAR);
				System.arraycopy(byteBlock, 0, decompress, ((ndx *
				 UNCOMPRESS_SIZE) / COMPRESS_SIZE), UNCOMPRESS_SIZE);
			}

			for (int i=0; i < size; i++)
				out.writeByte(decompress[i]);

			out.close ();
			in.close ();
		}

		catch (EOFException eof) {}

		catch (IOException e) {
			System.out.println ( "IO Exception =: " + e );
		}
	}
}
